src/extensions/default/ImageUrlCodeHints/camera.js

```diff
@@ -0,0 +1,187 @@
+                            navigator.mozGetUserMedia ||
+                            navigator.msGetUserMedia);
+
+
+    /**
+     * Dialog Buttons IDs
+     * @const {string}
+     */
+    var DIALOG_BTN_CANCEL = "cancel";
+    var DIALOG_BTN_OK = "ok";
+    
+    /**
+     * Dialog Buttons Class Names
+     * @const {string}
+     */
+    var DIALOG_BTN_CLASS_PRIMARY = "primary";
+    var DIALOG_BTN_CLASS_LEFT = "left";    
+
+    // Based on http://stackoverflow.com/questions/21797299/convert-base64-string-to-arraybuffer
+    function base64ToBuffer(base64) {
+        // This code will have to change to use Filer's Buffer api
+        var bin = window.atob(base64);
+        var len = bin.length;
+        var bytes = new window.Uint8Array(len);
+        for(var i = 0; i < len; i++) {
+            bytes[i] = bin.charCodeAt(i);
+        }
+
+        return bytes.buffer;
+    }
+
+    function streamVideo() {
+        if (streaming) {
+            return;
+        }
+
+        height = video.videoHeight / (video.videoWidth/width);
+
+        // Firefox currently has a bug where the height can't be read from
+        // the video, so we will make assumptions if this happens.
+        if (isNaN(height)) {
+            height = width / (4/3);
+        }
+
+        video.setAttribute('width', width);
+        video.setAttribute('height', height);
+        canvas.setAttribute('width', width);
+        canvas.setAttribute('height', height);
+        streaming = true;
+    }
+
+    function initWidget() {
+        var selfieContainer = $("#selfie-container");
+        $("#selfie-allow-access").remove();
+        selfieContainer.prepend(selfieWidgetHTML);
+        video = document.getElementById("selfie-video");
+        photo = document.getElementById("selfie-photo");
+        snap = document.getElementById("selfie-snap");
+        canvas = document.getElementById("selfie-canvas");
+        use = document.getElementById("selfie-use");
+        use.style.display = "initial";
+    }
+
+    function playVideo(stream) {
+        initWidget();
+
+        video.src = vendorURL.createObjectURL(stream);
+
+        video.addEventListener("canplay", streamVideo, false);
+        video.addEventListener("pause", function() {
+            vendorURL.revokeObjectURL(video.src);
+
+            if(stream) {
+                stream.stop();
+                stream = null;
+                streaming = false;
+            }
+        });
+        snap.addEventListener("click", function(ev) {
+            ev.preventDefault();
+            snapPhoto(deferred, filePath);
+        });
+
+        video.play();
+    }
+
+    function snapPhoto() {
+        var context = canvas.getContext("2d");
+        var data;
+
+        function savePhoto() {
+            if(!data) {
+                return deferred.reject();
+            }
+
+            var bin = /^data:image\/png;base64,(.+)/.exec(data)[1];
+            save(deferred, filePath, base64ToBuffer(bin));
```

Comment:
There's some weirdness in this file, like here where `deferred` and `filePath` are actually global and don't need to be passed to `save()`.